fix: notify networkmanager and networkobject not allowed

NoelStephensUnity opened the pull request:
Adding a GameObject with a NetworkObject component as a child to the GameObject that is assigned the NetworkManager component will cause a soft synchronization error. It is advised to avoid assigning NetworkObjects to the parent GameObject with the NetworkManager component or to any child GameObject of the parent GameObject. This only applies to a GameObject with the NetworkManager component assigned to it.

Users should be notified of this while within the editor and should not allow them to proceed with any configuration as described above.
[MTT-1598](https://jira.unity3d.com/browse/MTT-1598)
[MTT-1634](https://jira.unity3d.com/browse/MTT-1634)

## Changelog

### com.unity.netcode.gameobjects
- Fixed lack of notification that NetworkManager and NetworkObject cannot be added to the same GameObject with in-editor notifications

## Testing and Documentation
* No tests have been added (editor only update)

## Notification Displayed:
![image](https://user-images.githubusercontent.com/73188597/156858583-0c8e7715-80a2-4efb-87f3-30dbddb4db9b.png)

## NetworkObject Removed
![image](https://user-images.githubusercontent.com/73188597/156858621-46ae1631-ac8a-4ef7-8b53-e7076fe69ad0.png)

NoelStephensUnity committed:
fix

notify the user when they try to add a NetworkObject to a GameObject that already has a NetworkManager assigned to it.

NoelStephensUnity committed:
fix

extending the INetworkManagerHelper interface.

NoelStephensUnity committed:
update and style

Renamed the method being invoked to NotifyUserNetworkObjectRemoved in order to be more explicit in what was happening.
Updated comments

NoelStephensUnity committed:
Update CHANGELOG.md

NoelStephensUnity committed:
test and update

adding a test to validate that checking for a NetworkObject, removing the NetworkObject, and notifying the user that a NetworkManager cannot exist if a NetworkObject resides on the GameObject.  This also now checks children containing a NetworkObject.  It does not need to check for a parent as this is checked after to invoking NetworkManagerCheckForParent.

mattwalsh-unity commented on com.unity.netcode.gameobjects/Editor/NetworkManagerHelper.cs:
```diff
@@ -45,12 +45,57 @@ private static void EditorApplication_playModeStateChanged(PlayModeStateChange p
+            if (!EditorApplication.isUpdating)
+            {
+                Object.DestroyImmediate(networkObject);
+                var message = $"A {nameof(GameObject)} cannot have both a {nameof(NetworkManager)} and {nameof(NetworkObject)} assigned to it.";
```
This is cool, but what we don't tell users here is that you not only can't have a GO that has a NO and a NM, but you also cannot have children of a NM that have NOs.  I would like the message to convey that, else people can have a setup that matches your error message but is still broken.

Bonus points if you flag the "children" scenario and then have a dedicated message for that one (vs. one catchall message)

NoelStephensUnity replied:
Didn't need to add additional code for the child checking.  Just adjusted the language of the message.

mattwalsh-unity commented on com.unity.netcode.gameobjects/Tests/Editor/NetworkManagerConfigurationTests.cs:
```diff
@@ -34,5 +34,41 @@ public void NestedNetworkManagerCheck()
+        }
+
+        [Test]
+        public void NetworkObjectNotAllowed([Values] NetworkObjectPlacement networkObjectPlacement)
```
Nit, but I'd do "TestDisallowNetworkObjectsUnderNetworkManagers"

NoelStephensUnity replied:
This is the method:  NetworkManagerConfigurationTests.NetworkObjectNotAllowed
Do we really want to add another NetworkManager reference to this?
It would then be:
NetworkManagerConfigurationTests.TestDisallowNetworkObjectsUnderNetworkManagers
That seems a bit redundant?

mattwalsh-unity replied:
Ok, I'm sold

mattwalsh-unity commented on com.unity.netcode.gameobjects/Tests/Editor/NetworkManagerConfigurationTests.cs:
```diff
@@ -34,5 +34,41 @@ public void NestedNetworkManagerCheck()
+            var networkManager = gameObject.AddComponent<NetworkManager>();
+
+            // The error message we should expect
+            var messageToCheck = $"A {nameof(GameObject)} cannot have both a {nameof(NetworkManager)} and {nameof(NetworkObject)} assigned to it.";
```
IIRC we have banned the use of "checking for specific error messages", because it's so easy to forget that if you fix a message in one spot then the tests won't pass.

NoelStephensUnity replied:
Added method to return the error message.

NoelStephensUnity committed:
update

making a method to return the error message.
adjusting the error message to include children in the description.
adding comments for clarity.

mattwalsh-unity commented on com.unity.netcode.gameobjects/Tests/Editor/NetworkManagerConfigurationTests.cs:
```diff
@@ -34,5 +34,39 @@ public void NestedNetworkManagerCheck()
+            var networkManager = gameObject.AddComponent<NetworkManager>();
+
+            // Trap for the error message generated when a NetworkObject is discovered on the same GameObject or any children under it
+            LogAssert.Expect(LogType.Error, NetworkManagerHelper.Singleton.NetworkManagerAndNetworkObjectNotAllowedMessage());
```
since you have code that says “if I add something that can’t live under the NM, emit the message and delete it”, in addition to checking the message can you also verify that it has been removed?

NoelStephensUnity committed:
update

Adding a validation check to assure the NetworkObject has been removed.

NoelStephensUnity committed:
style

whitespace fix

NoelStephensUnity committed:
Merge branch 'develop' into fix/notify-networkmanager-and-networkobject-not-allowed
